src/main/java/com/hivemq/persistence/InMemorySingleWriter.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2019-present HiveMQ GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hivemq.persistence;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hivemq.configuration.service.InternalConfigurations;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import org.jctools.queues.MpscUnboundedArrayQueue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * @author Daniel Krüger
+ * <p>
+ * This SingleWriterService implementation does not use ExecutorThreadpools, but a trampoline approach like RxJava Schedulers.trampoline()
+ * The advantage is that there are less thread switches (resulting in context switches).
+ * The requirement is that no submitted task is blocking (that is only true for in-memory persistences)
+ */
+public class InMemorySingleWriter implements SingleWriterService {
+
+    private static final @NotNull Logger log = LoggerFactory.getLogger(SingleWriterServiceImpl.class);
+
+    private static final int AMOUNT_OF_PRODUCERS = 5;
+    private static final int RETAINED_MESSAGE_QUEUE_INDEX = 0;
+    private static final int CLIENT_SESSION_QUEUE_INDEX = 1;
+    private static final int SUBSCRIPTION_QUEUE_INDEX = 2;
+    private static final int QUEUED_MESSAGES_QUEUE_INDEX = 3;
+    private static final int ATTRIBUTE_STORE_QUEUE_INDEX = 4;
+
+    private final @NotNull InMemoryProducerQueues @NotNull [] producers = new InMemoryProducerQueues[AMOUNT_OF_PRODUCERS];
+    private final @NotNull InMemoryProducerQueues callbackProducerQueue;
+
+    public final @NotNull MpscUnboundedArrayQueue<Runnable> @NotNull [] queues;
+    public final @NotNull AtomicInteger @NotNull [] wips;
```

Comment:
queues and wips must be part of the InMemoryProducerQueues, not the single writer